ai_edge_torch/generative/layers/experimental/kv_cache.py

```diff
@@ -0,0 +1,178 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+# Cache management utilities.
+
+from dataclasses import dataclass
+from typing import List, Tuple
+
+import torch
+import torch.utils._pytree as pytree
+
+from ai_edge_torch import hlfb
+from ai_edge_torch.generative.layers import model_config
+
+
+@dataclass
+class KVCacheEntry:
+  """A single cache entry include K and V caches.
+
+  Returns:
+      _type_: _description_
+  """
+
+  k_cache: torch.Tensor
+  v_cache: torch.Tensor
+
+  @classmethod
+  def from_model_config(
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "KVCacheEntry":
+    shape = (
+        1,
+        config.kv_cache_max,
+        config.attn_config.num_query_groups,
+        config.head_dim,
+    )
+    k = torch.zeros(shape, dtype=dtype, device=device)
+    v = torch.zeros(shape, dtype=dtype, device=device)
+    obj = cls(k_cache=k, v_cache=v)
+    return obj
+
+
+@dataclass
+class KVCache:
+  """A utility class for holding KV cach entries per layer."""
+
+  caches: Tuple[KVCacheEntry]
+
+  @classmethod
+  def from_model_config(
+      cls,
+      config: model_config.ModelConfig,
+      dtype: torch.dtype = torch.float32,
+      device: torch.device = None,
+  ) -> "KVCache":
+    """Build an instance of the class based on model config.
+    Args:
+        config (ModelConfig): Model config used for building the cache.
+        dtype (torch.dtype, optional): The data type of the cache tensor.
+          Defaults to torch.float32.
+        device (torch.device, optional): The device placement of the cache
+          tensors. Defaults to None.
+    Returns:
+        KVCache: The created cache object.
+    """
+    caches = [
+        KVCacheEntry.from_model_config(config, dtype, device)
+        for _ in range(config.num_layers)
+    ]
+    obj = cls(caches=tuple(caches))
+    return obj
+
+
+def _flatten_kvc(kvc: KVCache) -> Tuple[List, List]:
+  flattened = []
+  flat_names = []
+  none_names = []
+  for i, kv_entry in enumerate(kvc.caches):
+    flattened.append(kv_entry.k_cache)
+    flat_names.append(f"k_{i}")
+    flattened.append(kv_entry.v_cache)
+    flat_names.append(f"v_{i}")
+  return flattened, [flat_names, none_names]
+
+
+def _flatten_kvc_with_keys(kvc: KVCache) -> Tuple[List, List]:
+  flattened, (flat_names, none_names) = _flatten_kvc(kvc)
```

Comment:
At first I thought I understood this, but now I realize I don't. Why is `none_names` created a returned? The function implementation doesn't use it and the caller who gets it back doesn't use it.

Reply:
It's the pytorch convention. They define `context` as list of names and non-names.